File: game/IdleMarket/Assets/Scripts/Equipments/EquipmentHelper.cs
using UnityEngine;

public static class EquipmentHelper
{
    public static StatType GetMainStat(EquipmentType type)
    {
        return type switch
        {
            EquipmentType.Sword => StatType.Attack,
            EquipmentType.Armor => StatType.Defense,
            EquipmentType.Helmet => StatType.Health,
            EquipmentType.Boots => StatType.Speed,
            _ => throw new System.ArgumentException($"Tipo de equipamento inválido: {type}"),
        };
    }

    public static StatRange GetMainStatRange(int rarity) => rarity switch
    {
        1 => new StatRange(5f, 10f),
        2 => new StatRange(10f, 20f),
        3 => new StatRange(20f, 30f),
        4 => new StatRange(30f, 40f),
        5 => new StatRange(40f, 50f),
        _ => throw new System.ArgumentException($"Raridade inválida: {rarity}"),
    };

    public static StatRange GetSubStatRange(StatType statType) => statType switch
    {
        StatType.Health or StatType.Attack or StatType.Defense => new StatRange(10f, 20f),
        StatType.CriticalChance => new StatRange(5f, 15f),
        StatType.CriticalDamage => new StatRange(10f, 30f),
        _ => throw new System.ArgumentException($"Tipo de estatística inválido: {statType}"),
    }; 
}

[System.Serializable]
public struct StatRange
{
    public float min; 
    public float max;

    public StatRange(float min, float max)
    {
        this.min = min;
        this.max = max;
    }

    public float Roll()
    {
        float value = Random.Range(min, max);
        return Mathf.Round(value * 10f) / 10f;
    }

    public float Normalize(float value)
    {
        return Mathf.InverseLerp(min, max, value);
    }
}

File: backend/src/game/equipmentHelper.ts
import { EquipmentType, StatType } from "../../generated/prisma/enums.js"
import { rangeFloat } from "./randomHelper.js"

export class StatRange {
    constructor(public min: number, public max: number) {}

    roll(): number {
        const value = rangeFloat(this.min, this.max)
        return Math.round(value * 10) / 10
    }

    normalize(value: number): number {
        const t = (value - this.min) / (this.max - this.min)
        return Math.min(1, Math.max(0, t))
    }
}

export function getMainStat(type: EquipmentType): StatType {
    switch (type){
        case EquipmentType.Sword: return StatType.Attack;
        case EquipmentType.Armor: return StatType.Defense;
        case EquipmentType.Helmet: return StatType.Health;
        case EquipmentType.Boots: return StatType.Speed;
        default: throw new Error(`tipo de equipamento inválida: ${type}`)
    }
}

export function getMainStatRange(rarity: number): StatRange {
    switch (rarity) {
        case 1: return new StatRange(5, 10);
        case 2: return new StatRange(10, 20);
        case 3: return new StatRange(20, 30);
        case 4: return new StatRange(30, 40);
        case 5: return new StatRange(40, 50);
        default: throw new Error(`Raridade inválida: ${rarity}`);
    }
}

export function getSubStatRange(statType: StatType): StatRange {
  switch (statType) {
    case StatType.Health:
    case StatType.Attack:
    case StatType.Defense:
      return new StatRange(10, 20);
    case StatType.CriticalChance:
      return new StatRange(5, 15);
    case StatType.CriticalDamage:
      return new StatRange(10, 30);
    default:
      throw new Error(`Tipo de status inválido: ${statType}`);
  }
}